internal/olm/operator/registry/operator_installer.go

```diff
@@ -34,12 +33,13 @@ import (
 )
 
 type OperatorInstaller struct {
-	CatalogSourceName string
-	PackageName       string
-	StartingCSV       string
-	Channel           string
-	InstallMode       operator.InstallMode
-	CatalogCreator    CatalogCreator
+	CatalogSourceName     string
+	PackageName           string
+	StartingCSV           string
+	Channel               string
+	InstallMode           operator.InstallMode
+	CatalogCreator        CatalogCreator
+	SupportedInstallModes sets.String
```

Comment:
Since we should always deduce the supported install modes from the CSV, we don't want to allow it to be defined by callers, right?

Do we need `SupportedInstallModes` to be exported?

Reply:
One downside to internalizing install mode lookup is having to load the bundle twice, once in `Install.Setup()` and again in `OperatorInstaller.InstallOperator()`.

Reply:
@estroz That's a great point. We can't do bundle loading at all in `OperatorInstaller.InstallOperator()` because that code is not packaging-aware.

Instead, we would need to use the packagemanifests API on-cluster to lookup the CSV. However, I don't think that API supports querying for a particular CSV version (only head of a channel), so we can't reliably determine the supported install modes for a particular CSV.

Reply:
So given that, my thinking is to leave `SupportedInstallModes` as an exported field and set it appropriately in the `bundle` and `packagemanifests` specific setup.

And we should check with OLM to see if it is on the roadmap to support specific version queries with the packagemanifests API add a TODO comment to move the logic to `OperatorInstaller` once that exists.

Reply:
> One downside to internalizing install mode lookup is having to load the bundle twice, once in `Install.Setup()` and again in `OperatorInstaller.InstallOperator()`.
That's exactly why I didn't do it. And passing the CSV around felt really weird. So I opted for the sets.String.

Reply:
I moved `GetSupportedInstallModes` to `install_mode.go` and exported it. Seemed to make more sense to put it there since it's just a utility to transform `InstallMode`s to a Set.

Reply:
Here's an interesting idea. What if instead of putting `InstallMode` and `SupportedInstallModes` fields on the `OperatorInstaller`, we actually just specify the `OperatorGroup` we want to be created.

If `OperatorGroup == nil`, `OperatorInstaller` would:
1. If no operator group exists in the namespace, error out.
2. Else, skip operator group creation

If `OperatorGroup` is not nil, `OperatorInstaller` would:
1. Check if an operator group already exists, and if so, error out.
2. Else, create the operator group

Our calling code would know all of this in advance, so it _shouldn't_ ever error, except for a race condition where another process or human changes the state in the middle.

This would help us keep the install mode parsing, compatibility checking, and conversion to target namespaces all together in one centralized place.

Assuming we've already validated the install mode flag, I could foresee a function signature like:
```go
func (i InstallMode) ToOperatorGroup(cfg operator.Configuration, csv v1alpha1.ClusterServiceVersion) (*v1.OperatorGroup, error)
```
in which we intersect `InstallMode.InstallModeType` with `csv.spec.installModes` and get the operator group(s) from the namespace, check existing for compatibility (and return a nil operator group or an error), or if none exists, build a new one and return it.